warehouse/query-core/src/main/java/datawave/query/index/lookup/BaseIndexStream.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * A reference to the underlying {@link datawave.query.tables.RangeStreamScanner} is required for seeking.
+ *
+ * Note that the BaseIndexStream does not implement the {@link IndexStream#seek(String)} method. Inheriting classes are responsible for determining the correct
+ * implementation.
+ */
+public abstract class BaseIndexStream implements IndexStream {
+    
+    protected RangeStreamScanner rangeStreamScanner;
+    
+    protected EntryParser entryParser;
+    
+    protected JexlNode node;
+    
+    protected StreamContext context;
+    
+    protected IndexStream debugDelegate;
+    
+    // variables to support the PeekingIterator interface
+    protected Iterator<Tuple2<String,IndexInfo>> backingIter;
+    protected Tuple2<String,IndexInfo> peekedElement;
+    protected boolean hasPeeked = false;
+    
+    public BaseIndexStream(RangeStreamScanner rangeStreamScanner, EntryParser entryParser, JexlNode node, StreamContext context, IndexStream debugDelegate) {
+        this.rangeStreamScanner = Preconditions.checkNotNull(rangeStreamScanner);
+        this.entryParser = Preconditions.checkNotNull(entryParser);
+        this.node = node;
+        this.backingIter = Iterators.transform(this.rangeStreamScanner, this.entryParser);
+        this.context = context;
+        this.debugDelegate = debugDelegate;
+    }
+    
+    public BaseIndexStream(Iterator<Tuple2<String,IndexInfo>> iterator, JexlNode node, StreamContext context, IndexStream debugDelegate) {
+        this.rangeStreamScanner = null;
+        this.entryParser = null;
+        this.node = node;
+        this.backingIter = Preconditions.checkNotNull(iterator);
+        this.context = context;
+        this.debugDelegate = debugDelegate;
+    }
+    
+    // Empty constructor used by the Union and Intersection classes.
+    public BaseIndexStream() {
+        
+    }
+    
+    /**
+     * Reset the backing iterator after a seek. State must stay in sync with changes to the RangeStreamScanner.
+     */
+    public void resetBackingIterator() {
+        if (rangeStreamScanner != null && entryParser != null) {
+            this.peekedElement = null;
+            this.hasPeeked = false;
+            this.backingIter = Iterators.transform(this.rangeStreamScanner, this.entryParser);
+        }
+    }
+    
+    @Override
+    public boolean hasNext() {
+        return (hasPeeked && peekedElement != null) || backingIter.hasNext();
+    }
+    
+    @Override
+    public Tuple2<String,IndexInfo> peek() {
+        if (!hasPeeked) {
+            if (backingIter.hasNext()) {
+                peekedElement = backingIter.next();
+            } else {
+                peekedElement = null;
+            }
+            hasPeeked = true;
+        }
+        return peekedElement;
+    }
+    
+    @Override
+    public Tuple2<String,IndexInfo> next() {
+        if (!hasPeeked) {
+            return backingIter.next();
```

Comment:
do we need a hasNext call here? Perhaps instead of this block just call peek() to avoid the duplication?

Reply:
I think as this is it conforms to the contract. If you haven't called hasNext() before calling next() you may get a NoSuchMethodException

Reply:
Strictly speaking, no. This method still conforms to the contract of the PeekingIterator. There is a way you could force this to break, but to do so you would also be violating the method contract of an iterator (by calling next() before calling hasNext()).